### PR TITLE
add dns document

### PR DESCRIPTION
**Pull request by junxiawang**
add dns document ,@tingtli please help review  and give comments
Link:
http://10.4.30.1/install/html/advanced/dns/index.html

### CODE REVIEW

File: docs/source/advanced/dns/cluster_name_resolution.rst [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+
+    127.0.0.1  localhost localhost.localdomain
+    50.1.2.3  mgmtnode-public mgmtnode-public.cluster.com
+    10.0.0.100  mgmtnode mgmtnode.cluster.com
+    10.0.0.1  node1 node1.cluster.com
+    10.0.0.2  node2 node2.cluster.com
+
+Verify that your **/etc/hosts** file contains entries for all of your management node interfaces. Manually add any that are missing.
+
+If your node names and IP addresses follow a regular pattern, you can easily populate **/etc/hosts** by putting a regular expression in the xCAT hosts table and then running makehosts. To do this, you need to first create an initial definition of the nodes in the database, if you haven't done that already: ::
```
Comment (penguhyang) [inline docs/source/advanced/dns/cluster_name_resolution.rst]: @junxiawang I found some words and syntax may need to correct. 
How about change then running makehosts to then running ``makehosts``

File: docs/source/advanced/dns/cluster_name_resolution.rst [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+
+    makehosts compute
+
+and the following entries will be added to **/etc/hosts**: ::
+
+    10.0.0.1 node01 node01.cluster.com
+    10.0.0.2 node02 node02.cluster.com
+    10.0.0.3 node03 node03.cluster.com
+
+This information is used by the ``makehosts`` command to add the additional interface hostnames etc. to the **/etc/hosts** file. It is also used by xCAT adapter configuration postscripts to automatically configure the additional network interfaces on the node. See the section Cluster_Name_Resolution/#specifying-additional-network-interfaces-for-cluster-nodes for more information.
```
Comment (penguhyang) [inline docs/source/advanced/dns/cluster_name_resolution.rst]: Cluster_Name_Resolution/#specifying-additional-network-interfaces-for-cluster-node this link didn't work normally.
FYI, http://xcat-docs.readthedocs.org/en/latest/developers/guides/docs/doc_guidelines.html#hyperlinks-internal-links-external-links

File: docs/source/advanced/dns/cluster_name_resolution.rst [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+
+        chdef -t site nameservers=<external dns IP>
+ 
+   * Set the correct information of external dns into the /etc/resolv.conf on your xCAT MN.
+
+The **domain** and **nameservers** values must be set correctly in **/etc/resolv.conf**. Which should have the same values with the ones your set in the site table.
+
+   *  Manually set up your external dns server with correct named.conf and correct zone files
+
+   *  Add the TSIG to the named.conf of your external dns for makedns command to update external dns ::
```
Comment (penguhyang) [inline docs/source/advanced/dns/cluster_name_resolution.rst]: maybe need to add a dot(.) before ：： to set the indent.

File: docs/source/advanced/dns/cluster_name_resolution.rst [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+
+      tabdump -w key==omapi passwd
+      get the key like "omapi","xcat_key","MFRCeHJybnJxeVBNaE1YT1BFTFJzN2JuREFMeEMwU0U=",,,,
+      Add it to your named.conf
+      key xcat_key {
+           algorithm hmac-md5;
+           secret "MFRCeHJybnJxeVBNaE1YT1BFTFJzN2JuREFMeEMwU0U=";
+      };
+   * Then change each zone to make your zones to allow this key to update.
+      zone "1.168.192.IN-ADDR.ARPA." in {
```
Comment (penguhyang) [inline docs/source/advanced/dns/cluster_name_resolution.rst]: I think we should need to add an empty line before this line.

File: docs/source/advanced/dns/cluster_name_resolution.rst [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+
+    chdef -t node -o compute02 niccustomscripts.eth1=cfgeth niccustomscripts.eth2='cfgeth|cfgeth'
+
+In this example **"cfgeth"** is the name of an adapter configuration postscript to be used to configure the interface.
+
+nicnetworks example: ::
+
+    chdef -t node -o compute02 nicnetworks.eth1=clstrnet11 nicnetworks.eth2='clstrnet80|clstrnet-lab'
+
+In this example we are saying that the IP address of "eth0" (ie. compute02-eth1 -> 11.10.1.2) is part of the xCAT network named **"clstrnet11"**. **"compute02-eth2"** is in network **"clstrnet80"** and **"compute02-eth2-lab"** is in **"clstrnet-lab"**.
```
Comment (penguhyang) [inline docs/source/advanced/dns/cluster_name_resolution.rst]: I think we should correct eth0 to eth1 in this sentence. (In this example we are saying that the IP address of "eth0")

File: docs/source/advanced/dns/cluster_name_resolution.rst [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The **xCAT *defs** commands **(mkdef, chdef, and lsdef)** may be used to manage the additional NIC information in the xCAT database.
+
+When using the these commands the expanded nic* attribute format will always be used.
+
+Expanded format for nic* attributes
+''''''''''''''''''''''''''''''''''''
+
+The expanded format will be the nics attribute name and the nic name, separated by a "." (dot).(ie. <nic attr="" name="">.<nic name=""> )
```
Comment (penguhyang) [inline docs/source/advanced/dns/cluster_name_resolution.rst]: How about change (ie. <nic attr="" name="">.<nic name=""> ) to (ie. <nic attribute name="">.<nic name=""> )

File: docs/source/advanced/dns/cluster_name_resolution.rst [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+   * Then change each zone to make your zones to allow this key to update.
+      zone "1.168.192.IN-ADDR.ARPA." in {
+           type master;
+           allow-update {
+                   key xcat_key;
+           };
+           file "db.192.168.1";
+      };
+
+    To update the name resolution entries from /etc/hosts or hosts table of xCAT MN to external dns
```
Comment (penguhyang) [inline docs/source/advanced/dns/cluster_name_resolution.rst]: maybe need to add a * in front of this sentence.

File: docs/source/advanced/index.rst [hunk 8]
```diff
@@ -21,3 +21,4 @@ Advanced Topics
    sysclone/index.rst
    webservices/index.rst
    zones/index.rst
+   dns/index.rst
```
Comment (whowutwut) [inline docs/source/advanced/index.rst]: This is added to the wrong place, the documentation has been alphabetized... 

I don't think we should use ``dns`` as the sub directory if we are going to call the page ``Cluster Name Resolution``.

File: docs/source/advanced/dns/index.rst [hunk 9]
```diff
@@ -0,0 +1,7 @@
+Cluster Name Resolution
```
Comment (whowutwut) [inline docs/source/advanced/dns/index.rst]: I think the title should be "Domain Name Resolution"

**Comment by whowutwut** [comment]
@junxiawang  you can assign @tingtli  or @penguhyang, we should make sure there are owners for the pull request, if possible.

**Comment by junxiawang** [comment]
@whowutwut owner is @tingtli .I have told here.

**Comment by junxiawang** [comment]
@whowutwut 
Document for  dns 's review owner is @tingtli.I told her and she will review it later.

**Comment by tingtli** [comment]
@whowutwut @penguhyang  I am very busy with end of cycle regression. Probably delay the document review. Please help review this document.  And I will reivew this when I have time.

File: docs/source/advanced/domain_name_resolution/domain_name_resolution.rst [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+There are many different ways to configure name resolution in your cluster. Four of the common choices will be described in this section (look for Option #: headings): ::
+
+    1.In a basic (non-hierarchical) cluster, point all nodes to a DNS server running on the management node. This is the most common setup.
+    2.In a basic (non-hierarchical) cluster, point all nodes to an external DNS running at your site. This requires that all of your nodes have network connectivity to your site.
+    3.In a hierarchical cluster, point all compute nodes to their service node. Make service node as the compute nodes' DNS server.
+    4.Don't use DNS, just distribute the /etc/hosts file to every node. If you choose this you will have to distribute new versions of the /etc/hosts file to all the cluster nodes whenever you add new nodes to the cluster, and you will have to specify site.master and all other server attributes in the database as IP addresses.
+
+But before any of those options are chosen, there are some things that must be done for all of the options.
+
+Set the site domain Attribute
```
Comment (zet809) [inline docs/source/advanced/domain_name_resolution/domain_name_resolution.rst]: Set domain name for the cluster

File: docs/source/advanced/domain_name_resolution/domain_name_resolution.rst [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+First set the domain attribute in the xCAT site table to the hostname domain you want to use for your cluster nodes: ::
```
Comment (zet809) [inline docs/source/advanced/domain_name_resolution/domain_name_resolution.rst]: To set domain name for the cluster, use: ::

File: docs/source/advanced/domain_name_resolution/domain_name_resolution.rst [hunk 12]
```diff
@@ -0,0 +1,499 @@
...
+If you want to use a different hostname domain or a different set of nameservers for nodes that are on a particular network, set those attributes in the corresponding network object: ::
+
+    mkdef -t network clusternet domain=app.cluster.com nameservers=1.2.3.4
+
+Populate the /etc/hosts File
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+All of the management node interfaces and all of the nodes need to be added to the **/etc/hosts** file on the xCAT management node (whether you are using the DNS option or not). You can either edit the **/etc/hosts** file by hand, or use ``makehosts``.
+
+If you edit the file by hand, it should look similar to: ::
```
Comment (zet809) [inline docs/source/advanced/domain_name_resolution/domain_name_resolution.rst]: Edit /etc/hosts manually or with 'makehost', the file are similar, right?

File: docs/source/advanced/domain_name_resolution/domain_name_resolution.rst [hunk 13]
```diff
@@ -0,0 +1,499 @@
...
+
+    makehosts compute
+
+and the following entries will be added to **/etc/hosts**: ::
+
+    10.0.0.1 node01 node01.cluster.com
+    10.0.0.2 node02 node02.cluster.com
+    10.0.0.3 node03 node03.cluster.com
+
+This information is used by the ``makehosts`` command to add the additional interface hostnames etc. to the **/etc/hosts** file. It is also used by xCAT adapter configuration postscripts to automatically configure the additional network interfaces on the node. See the section (refer to :ref:`specifying_additional_network_interfaces_for_cluster_nodes` ).
```
Comment (zet809) [inline docs/source/advanced/domain_name_resolution/domain_name_resolution.rst]: The information is created by "makehosts" not used by it, right?

File: docs/source/advanced/domain_name_resolution/domain_name_resolution.rst [hunk 14]
```diff
@@ -0,0 +1,499 @@
...
+Note that it is a convention of xCAT that for Linux systems the short hostname is the primary hostname for the node, and the long hostname is an alias. To have the long hostname be the primary hostname, you can use the -l option on the makehosts command.
+
+Preparing for Using a DNS
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you are choosing any of the options for using DNS, follow these steps:
+
+NOTE: This documentation only applies to the xCAT makedns command using the ddns.pm plugin. The ddns.pm plugin is based on named9/bind9, and can not support named8/bind8 due to syntax difference. 
+
+    * Set the **nameservers** and **forwarders** attributes in the xCAT site table. The **nameservers** attribute identifies the DNS server hostname/ip that the nodes point to in their **/etc/resolv.conf** files. The forwarders attribute are the DNS server's ip that can resolve external hostnames. If you are running a DNS on the xCAT MN, it will use the forwarders DNS server to resolve any hostnames it can't.
```
Comment (zet809) [inline docs/source/advanced/domain_name_resolution/domain_name_resolution.rst]: "Set the **nameservers** and **forwarders** attributes in the xCAT site table" is the subtitle, right?

**Pull request merged**